README.md
# MMM-octoprint

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

![screenshot](.github/example.png)

This module will use the Octoprint API to show your current print status on your MagicMirror²!

## Installation

In your terminal, go to your MagicMirror's Module folder:

```bash
cd ~/MagicMirror/modules
```

Clone this repository:

```bash
git clone https://github.com/btastic/MMM-octoprint
```

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-octoprint',
            position: 'top_right',
            config: {
                endpoint: "",
                apiKey: "",
                updateInterval: 60000,
                showThumbnail: false,
                showLayerProgress: false,
                thumbnailSize: 150,
                hideDataOnStandby: true
            }
        }
    ]
}
```

## Layer progress
To display the layer progress you have to install the [DisplayLayerProgress](https://plugins.octoprint.org/plugins/DisplayLayerProgress/) plugin. Please follow the [instructions](https://github.com/OllisGit/OctoPrint-DisplayLayerProgress/wiki/How-does-the-plugin-works) to set up your slicer accordingly or use the automatic detection of `;LAYER:<num>` during gcode upload.

## Thumbnails
To display a thumbnail you have to install the [Slicer Thumbnails](https://plugins.octoprint.org/plugins/prusaslicerthumbnails/) plugin.

## Customization
To change the color of the progress bar, you can override the default colors (green on gray) with css in your custom.css file.

You can also change the border radius of the progress bar, if you don't like rounded corners.

For example:
```css
.MMM-octoprint {
  --progress-color: crimson;
  --progress-background-color: gray;
  --border-radius: 0px;
}
```

## Configuration options

| Option               | Description
|--------------------- |-----------
| `endpoint`           | *Required* <br>Endpoint of the Octoprint API. For example: http://192.168.0.75<br><br>**Type:** `string`(url) 
| `apiKey`           | *Required* <br>API Key for Authorization. To create an application key, follow the instructions on [Octoprint docs](https://docs.octoprint.org/en/master/bundledplugins/appkeys.html)<br><br>**Type:** `string` 
| `updateInterval`     | *Optional* <br>The time interval between UI updates.<br><br>**Type:** `int`(milliseconds)<br>**Default:** 60000 milliseconds (60 seconds)
| `showThumbnail`      | *Optional* <br>Shows a thumbnail embedded in your gcode files. <br><br>**Only enable this, once you configured the thumbnails plugin and your slicer.**<br><br>**Type:** `bool` <br>**Default:** false
| `showLayerProgress`  | *Optional* <br>Shows the layer progress. <br><br> **Only enable this, once you configured the layer progress plugin and your slicer.**<br><br>**Type:** `bool` <br>**Default:** false
| `thumbnailSize`      | *Optional* <br>Changes with thumbnails size (width). Enter a integer value for an absolute value, or a relative size with % (e.g. "50%").<br><br>**Type:** `int\|string` <br>**Default:** 150
| `hideDataOnStandby`  | *Optional* <br>Hides printer info when the printer is operational.<br><br>**Type:** `bool` <br>**Default:** true

## Available localizations
- English (en)
- German (de)

CHANGELOG.md
# MMM-octoprint Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0]

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-octoprint
 *
 * By Ben Konsemüller
 * MIT Licensed.
 */

const Log = require("logger");
const NodeHelper = require("node_helper");
const fetch = require("fetch");
const moment = require("moment");

module.exports = NodeHelper.create({
  config: {},
  currentFile: null,
  async socketNotificationReceived(notification, payload) {
    if (notification === "CONFIG") {
      this.config = payload;

      if (this.fetchTimerId) {
        clearTimeout(this.fetchTimerId);
      }

      await this.fetchData();
    }
  },

  async fetchData() {
    const self = this;

    const printer_status = await this.fetchPrinterStatus();

    if (!printer_status) {
      return;
    }

    const job_status = await this.fetchPrinterJob();

    let thumbnail = null;
    let layer_information = null;

    if (this.config.showThumbnail) {
      thumbnail = await this.fetchThumbnail(job_status);
    }

    if (this.config.showLayerProgress) {
      layer_information = await this.fetchLayerInformation();
    }

    const eta = moment.utc(1000 * (job_status.progress.printTimeLeft)).format('HH[h] mm[m] ss[s]');

    this.sendSocketNotification("PRINTER_STATUS", { printer_status, job_status, eta, layer_information, thumbnail });

    this.fetchTimerId = setTimeout(async function () {
      await self.fetchData();
    }, this.config.updateInterval);
  },

  async fetchPrinterStatus() {
    const endpoint = this.config.endpoint + "/api/printer";

    try {
      const response = await fetch(endpoint, { headers: this.getHeaders() });
      const json = await response.json();

      return json;
    } catch (error) {
      Log.error(`${this.name} received an error: ${error}`);
      this.sendSocketNotification("HTTP_ERROR", {});

      return null;
    }
  },

  async fetchPrinterJob() {
    const endpoint = this.config.endpoint + "/api/job";

    try {
      const response = await fetch(endpoint, { headers: this.getHeaders() });
      const json = await response.json();

      return json;
    } catch (error) {
      Log.error(`${this.name} received an error: ${error}`);
      this.sendSocketNotification("HTTP_ERROR", {});

      return null;
    }
  },

  async fetchLayerInformation() {
    const endpoint = this.config.endpoint + "/plugin/DisplayLayerProgress/values";

    try {
      const response = await fetch(endpoint, { headers: this.getHeaders() });
      const json = await response.json();

      return json;
    } catch (error) {
      Log.error(`${this.name} received an error: ${error}`);
      this.sendSocketNotification("HTTP_ERROR", {});

      return null;
    }
  },

  async fetchThumbnail(job_status) {
    if (!job_status.job.file.name) {
      return;
    }

    const endpoint = this.config.endpoint + "/api/files/" + job_status.job.file.origin + "/" + job_status.job.file.name;

    try {
      const response = await fetch(endpoint, { headers: this.getHeaders() });
      const json = await response.json();

      if (!json.thumbnail) {
        json.thumbnail = "./modules/MMM-octoprint/img/no_thumbnail.png";
      } else {
        json.thumbnail = this.config.endpoint + "/" + json.thumbnail;
      }

      return json.thumbnail;
    } catch (error) {
      Log.error(`${this.name} received an error: ${error}`);
      this.sendSocketNotification("HTTP_ERROR", {});

      return null;
    }
  },

  getHeaders() {
    return {
      "Authorization": `Bearer ${this.config.apiKey}`
    };
  },
});


MMM-octoprint.js
/* global Module */

/* Magic Mirror
 * Module: MMM-octoprint
 *
 * By Ben Konsemüller
 * MIT Licensed.
 */

Module.register("MMM-octoprint", {
  defaults: {
    endpoint: "",
    apiKey: "",
    updateInterval: 60000,
    showThumbnail: false,
    showLayerProgress: false,
    thumbnailSize: 150,
    hideDataOnStandby: true
  },

  // Store the data in an object.
  displayData: {
    printer_status: null,
    eta: null,
    thumbnails: [],
  },

  available: true,
  loading: true,

  requiresVersion: "2.1.0", // Required version of MagicMirror

  start: function () {
    Log.info(`Starting module: ${this.name}`);
    this.sendSocketNotification("CONFIG", this.config);
  },

  socketNotificationReceived: function (notification, payload) {
    switch (notification) {
      case "PRINTER_STATUS":
        this.displayData = payload;
        this.available = true;
        this.loading = false;
        break;
      case "HTTP_ERROR":
        this.loading = false;
        this.available = false;
        break;
    }
    this.updateDom(this.config.animationSpeed);
  },

  getTemplate: function () {
    return "templates\\mmm-octoprint.njk";
  },

  getTemplateData() {
    const templateData = {
      loading: this.loading,
      config: this.config,
      printer_status: this.loading ? null : this.displayData.printer_status,
      job_status: this.loading ? null : this.displayData.job_status,
      eta: this.loading ? null : this.displayData.eta,
      thumbnail: this.loading ? null : this.displayData.thumbnail,
      layer_information: this.loading ? null : this.displayData.layer_information,
      available: this.available,
    };

    return templateData;
  },

  getScripts: function () {
    return [];
  },

  getStyles: function () {
    return [
      "MMM-octoprint.css",
    ];
  },

  getTranslations: function () {
    return {
      en: "translations/en.json",
      de: "translations/de.json",
    };
  },
});
